src/BSON/Value.c

```diff
@@ -0,0 +1,631 @@
+/*
+ * Copyright 2014-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <php.h>
+#include <ext/standard/base64.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_operators.h>
+#include <ext/standard/php_var.h>
+#include <zend_smart_str.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+#include "phongo_bson_encode.h"
+#include "BSON/Value_arginfo.h"
+#include "BSON/Binary.h"
+#include "BSON/DBPointer.h"
+#include "BSON/Decimal128.h"
+#include "BSON/Document.h"
+#include "BSON/Int64.h"
+#include "BSON/Javascript.h"
+#include "BSON/ObjectId.h"
+#include "BSON/PackedArray.h"
+#include "BSON/Regex.h"
+#include "BSON/Symbol.h"
+#include "BSON/Timestamp.h"
+#include "BSON/UTCDateTime.h"
+
+#define PHONGO_VALUE_TYPE_ISSER(name, type)                              \
+	static PHP_METHOD(MongoDB_BSON_Value, name)                          \
+	{                                                                    \
+		PHONGO_PARSE_PARAMETERS_NONE();                                  \
+		RETURN_BOOL(Z_VALUE_OBJ_P(getThis())->value.value_type == type); \
+	}
+
+#define PHONGO_VALUE_CHECK_TYPE(bson_value, expected_type)                                                       \
+	if (bson_value.value_type != expected_type) {                                                                \
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Unexpected BSON type %d", bson_value.value_type); \
+		return;                                                                                                  \
+	}
+
+zend_class_entry* php_phongo_value_ce;
+
+static HashTable* php_phongo_value_get_properties_hash(phongo_compat_object_handler_type* object, bool is_temp)
+{
+	php_phongo_value_t* intern;
+	HashTable*          props;
+
+	intern = Z_OBJ_VALUE(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_temp, intern, props, 1);
+
+	// TODO: Print value (somehow)
```

Comment:
I was thinking of including the type here, and we may just want to include the result of `bson_value_to_zval` here instead of displaying raw data.

Reply:
Reporting the BSON type along with the zeal makes sense, I'll implement that 👍